Voiceover pagedown

zarend opened the pull request:
Please see commit message for description.

crisbeto commented on src/material/datepicker/calendar-body.ts:
```diff
@@ -195,23 +195,29 @@ export class MatCalendarBody implements OnChanges, OnDestroy {
-          if (!movePreview) {
-            this._skipNextFocus = true;
+    setTimeout(() => {
+      this._ngZone.runOutsideAngular(() => {
```
Should the `runOutsideAngular` be outside of the `setTimeout`?

zarend replied:
yes, that's a good idea. I moved the timeout to be inside of the `onStable`. `The less things happening during the `timeOut` timer, hopeful the more consistent behavior we should get. It might take longer for `runOutsideAngular`, and `onStable` to happen depending on the situation.

crisbeto commented on src/material/datepicker/calendar-body.ts:
```diff
@@ -195,23 +195,29 @@ export class MatCalendarBody implements OnChanges, OnDestroy {
+        });
       });
-    });
+    }, 20);
```
Is there a reason for the number 20 specifically?

zarend replied:
When I tested this, it worked consistently with 10ms timeout, so I make it 20 to be safe.

crisbeto commented on src/material/datepicker/calendar-header.html:
```diff
@@ -3,7 +3,7 @@
             (click)="currentPeriodClicked()" [attr.aria-label]="periodButtonLabel"
             [attr.aria-describedby]="_buttonDescriptionId"
-            cdkAriaLive="polite">
+            [attr.aria-live]="'polite'">
```
You can write as just `aria-live="polite"`.

crisbeto commented on src/material/datepicker/calendar-header.html:
```diff
@@ -3,7 +3,7 @@
             (click)="currentPeriodClicked()" [attr.aria-label]="periodButtonLabel"
             [attr.aria-describedby]="_buttonDescriptionId"
-            cdkAriaLive="polite">
+            attr.aria-live="polite">
```
Should be `aria-live="polite"`.

zarend replied:
👍

zarend commented:
Changing this to a draft, as I need to fix many tests. Adding the 20ms timeout means many tests need to be updated to have a `tick(20)`. I'll change it to a real PR this this is ready.

jelbourn commented on src/material/datepicker/calendar-body.ts:
```diff
@@ -31,6 +31,8 @@ export type MatCalendarCellClassFunction<D> = (
   view: 'month' | 'year' | 'multi-year',
 ) => MatCalendarCellCssClasses;
 
+export const FOCUS_ACTIVE_CELL_DELAY = 20;
```
Add a comment that explains what this constant is and where the value came from (as detailed as possible)

jelbourn commented on src/material/datepicker/calendar-body.ts:
```diff
@@ -216,10 +218,14 @@ export class MatCalendarBody implements OnChanges, OnDestroy, AfterViewChecked {
   _focusActiveCell(movePreview = true) {
     this._ngZone.runOutsideAngular(() => {
-      this._ngZone.onStable.pipe(take(1)).subscribe(() => {
+      this._ngZone.onStable.pipe(take(1), delay(FOCUS_ACTIVE_CELL_DELAY)).subscribe(() => {
```
Add a comment here that explains the purpose of the `delay`

zarend replied:
There's an explanation for the `delay` a few lines above, which is right before the declaration of `_focusActiveCell` :)

jelbourn reviewed (commented):
Any thoughts on whether we could add a unit test here?

zarend commented:
@jelbourn and @crisbeto , I address the comments and cleaned up this PR. It's ready for your eyes again 👀 .

After further testing, I determined 0ms would be best for the timeout duration since longer could interfere with repeated keys and cause the entire page to scrolling while pressing and holding pagedown.

I added a code comment with more details on my testing methodology and how I determined 0ms would be best. The code comment is there so we can consider why it was set to 0ms if we ever want to change the timeout value later.

crisbeto reviewed (approved):
LGTM

mmalerba reviewed (approved):
LGTM
